File: src/js/cable_joints/createCablePaths.js
import Vector2 from './vector2.js';

import {
  CablePathComponent
} from './cable_joints_core.js';

export function createCablePaths(world, jointEntities = [], linkTypes = [], cw = [], spring_constant = 1e6, userStored = null) {
    const createdPathEntityIds = [];

    // Validate input array lengths relative to each other for a single conceptual path
    if (linkTypes.length !== jointEntities.length + 1) {
        console.warn("createCablePaths: linkTypes.length must be jointEntities.length + 1. Aborting.");
        return createdPathEntityIds;
    }
    if (cw.length !== linkTypes.length) {
        console.warn("createCablePaths: cw.length must be linkTypes.length. Aborting.");
        return createdPathEntityIds;
    }
    if (userStored && userStored.length !== linkTypes.length) {
        console.warn("createCablePaths: userStored.length must be linkTypes.length if provided. Aborting.");
        return createdPathEntityIds;
    }

    if (jointEntities.length === 0) {
        // This means linkTypes.length is 1 (due to validation above). A single link, no joints.
        // Create one CablePathComponent.
        const pathEntityId = world.createEntity();
        const pathComponent = new CablePathComponent(world, [], linkTypes, cw, spring_constant, userStored);
        world.addComponent(pathEntityId, pathComponent);
        createdPathEntityIds.push(pathEntityId);
        return createdPathEntityIds;
    }

    let currentPathJoints = [];
    let currentPathLinkTypes = [];
    let currentPathCw = [];
    let currentPathStored = userStored ? [] : null;

    // The first link type always starts the first path segment
    currentPathLinkTypes.push(linkTypes[0]);
    currentPathCw.push(cw[0]);
    if (userStored) {
        currentPathStored.push(userStored[0]);
    }

    for (let i = 0; i < jointEntities.length; i++) {
        const jointId = jointEntities[i];
        const linkTypeAfterJoint = linkTypes[i + 1];
        const cwAfterJoint = cw[i + 1];
        const storedAfterJoint = userStored ? userStored[i + 1] : null;

        currentPathJoints.push(jointId);
        currentPathLinkTypes.push(linkTypeAfterJoint);
        currentPathCw.push(cwAfterJoint);
        if (userStored) {
            currentPathStored.push(storedAfterJoint);
        }

        // Check if the link *after* the current joint is an 'attachment'
        // and if it's not the very last link in the overall path definition.
        // (i + 1) is the index of linkTypeAfterJoint in the original linkTypes array.
        // linkTypes.length - 1 is the index of the last link type in the original linkTypes array.
        if (linkTypeAfterJoint === 'attachment' && (i + 1) < (linkTypes.length - 1)) {
            // Finalize current path
            const pathEntityId = world.createEntity();
            const pathComponent = new CablePathComponent(world, currentPathJoints, currentPathLinkTypes, currentPathCw, spring_constant, currentPathStored);
            world.addComponent(pathEntityId, pathComponent);
            createdPathEntityIds.push(pathEntityId);

            // Start a new path
            currentPathJoints = [];
            // The 'attachment' link that caused the split starts the new path segment.
            currentPathLinkTypes = [linkTypeAfterJoint];
            currentPathCw = [cwAfterJoint];
            if (userStored) {
                currentPathStored = [storedAfterJoint];
            }
        }
    }

    // Add the last (or only remaining) path segment
    if (currentPathJoints.length > 0 || (createdPathEntityIds.length === 0 && linkTypes.length > 0)) {
        // The second condition (createdPathEntityIds.length === 0 && linkTypes.length > 0)
        // should not be strictly necessary due to the jointEntities.length === 0 check at the start,
        // but kept for robustness in case of unusual inputs that might bypass earlier checks
        // and result in no joints processed but still a valid single segment.
        const pathEntityId = world.createEntity();
        const pathComponent = new CablePathComponent(world, currentPathJoints, currentPathLinkTypes, currentPathCw, spring_constant, currentPathStored);
        world.addComponent(pathEntityId, pathComponent);
        createdPathEntityIds.push(pathEntityId);
    }

    return createdPathEntityIds;
}

File: src/python/cable_joints/create_cable_paths.py
import warnings
from typing import List, Optional, cast

from .ecs import World
from .cable_joints_components import create_cable_path_component

def create_cable_paths(
    world: World,
    joint_entities: Optional[List[int]] = None,
    link_types: Optional[List[str]] = None,
    cw: Optional[List[bool]] = None,
    spring_constant: float = 1e6,
    user_stored: Optional[List[Optional[float]]] = None,
    cable_half_width: float = 0.0
) -> List[int]:
    """
    Creates one or more CablePathComponent entities from a definition of joints and links.

    This function takes a series of joints and link properties and constructs
    CablePathComponent entities. It splits the series into multiple paths if it
    encounters an 'attachment' link type that is not at the very end of the path.
    This is a Python port of the createCablePaths.js function.

    Args:
        world: The ECS world object.
        joint_entities: A list of entity IDs for the joints in the path.
        link_types: A list of strings describing the type of each link.
                    len(link_types) must be len(joint_entities) + 1.
        cw: A list of booleans indicating clockwise preference for rolling links.
            len(cw) must be len(link_types).
        spring_constant: The spring constant for the cable paths.
        user_stored: Optional list of pre-calculated stored lengths for links.
                     If provided, len(user_stored) must be len(link_types).

    Returns:
        A list of entity IDs for the created CablePathComponent entities.
    """
    if joint_entities is None:
        joint_entities = []
    if link_types is None:
        link_types = []
    if cw is None:
        cw = []

    created_path_entity_ids: List[int] = []

    # Validate input array lengths
    if len(link_types) != len(joint_entities) + 1:
        warnings.warn("create_cable_paths: len(link_types) must be len(joint_entities) + 1. Aborting.", UserWarning)
        return created_path_entity_ids
    if len(cw) != len(link_types):
        warnings.warn("create_cable_paths: len(cw) must be len(link_types). Aborting.", UserWarning)
        return created_path_entity_ids
    if user_stored is not None and len(user_stored) != len(link_types):
        warnings.warn("create_cable_paths: len(user_stored) must be len(link_types) if provided. Aborting.", UserWarning)
        return created_path_entity_ids

    if not joint_entities:
        # This means link_types has length 1. A single link, no joints.
        path_entity_id = world.create_entity()
        path_component = create_cable_path_component(
            world,
            [],
            link_types,
            cw,
            spring_constant,
            user_stored,
            cable_half_width
        )
        world.add_component(path_entity_id, path_component)
        created_path_entity_ids.append(path_entity_id)
        return created_path_entity_ids

    current_path_joints: List[int] = []
    current_path_link_types: List[str] = []
    current_path_cw: List[bool] = []
    current_path_stored: Optional[List[Optional[float]]] = [] if user_stored is not None else None

    # The first link type always starts the first path segment
    current_path_link_types.append(link_types[0])
    current_path_cw.append(cw[0])
    if user_stored is not None:
        cast(list, current_path_stored).append(user_stored[0])

    for i, joint_id in enumerate(joint_entities):
        link_type_after_joint = link_types[i + 1]
        cw_after_joint = cw[i + 1]
        stored_after_joint = user_stored[i + 1] if user_stored is not None else None

        current_path_joints.append(joint_id)
        current_path_link_types.append(link_type_after_joint)
        current_path_cw.append(cw_after_joint)
        if user_stored is not None:
            cast(list, current_path_stored).append(stored_after_joint)

        # Check if the link *after* the current joint is an 'attachment'
        # and if it's not the very last link in the overall path definition.
        if link_type_after_joint == 'attachment' and (i + 1) < (len(link_types) - 1):
            # Finalize current path
            path_entity_id = world.create_entity()
            path_component = create_cable_path_component(
                world,
                current_path_joints,
                current_path_link_types,
                current_path_cw,
                spring_constant,
                current_path_stored,
                cable_half_width
            )
            world.add_component(path_entity_id, path_component)
            created_path_entity_ids.append(path_entity_id)

            # Start a new path
            current_path_joints = []
            # The 'attachment' link that caused the split starts the new path segment.
            current_path_link_types = [link_type_after_joint]
            current_path_cw = [cw_after_joint]
            if user_stored is not None:
                current_path_stored = [stored_after_joint]

    # Add the last (or only remaining) path segment
    if current_path_joints or (not created_path_entity_ids and link_types):
        path_entity_id = world.create_entity()
        path_component = create_cable_path_component(
            world,
            current_path_joints,
            current_path_link_types,
            current_path_cw,
            spring_constant,
            current_path_stored,
            cable_half_width
        )
        world.add_component(path_entity_id, path_component)
        created_path_entity_ids.append(path_entity_id)

    return created_path_entity_ids
